### PR TITLE
Fix broken matplotlib/minianalyses

### PR DESCRIPTION
**Pull request by JoranAngevaare**
# Changes due to matplotlib3.5
 - Passing vmin and vmax independent of lognorm results in errors
 - `plt.tight_layout()` doesn't like how the cbar is added in the records matrix

# Changes in minianlyses
Since we are using all kinds of funky tricks in the minianalyses I added some more tests there as small things like the changes from matplotlib 3.5 might have otherwise gone unnoticed.

# Bugfixes
 - fix requirements file
 - several scaffolding from the static event display
 - Unclear issues with the pulse plotting when no time range is specified
 - unclear tracebacks in `plot_energy_spectrum`
 - remove assert statements
 - Use constrained layout instead of tight layout
 - many more mini-analyses tests

### CODE REVIEW

**Comment by JoranAngevaare** [comment]
relates to https://github.com/XENONnT/base_environment/pull/790 and https://github.com/XENONnT/straxen/pull/805

File: straxen/analyses/bokeh_waveform_plot.py [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
-
-import numpy as np
 import numba
+import numpy as np
 import strax
 import straxen
-
-import warnings
+from straxen.analyses.holoviews_waveform_display import _hvdisp_plot_records_2d, hook, \
+    plot_record_polygons, get_records_matrix_in_window
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:10]:
```suggestion
from straxen.analyses.holoviews_waveform_display import _hvdisp_plot_records_2d, hook, plot_record_polygons, get_records_matrix_in_window  # noqa
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 2]
```diff
@@ -143,7 +141,8 @@ def event_display_interactive(events,
     if np.any(m_other_s2) and not only_main_peaks:
         # Now we have to add the positions of all the other S2 to the top pmt array
         # if not only main peaks.
-        fig_top, plot = plot_posS2s(peaks[m_other_s2], label='OS2s', fig=fig_top, s2_type_style_id=2)
+        fig_top, plot = plot_posS2s(peaks[m_other_s2], label='OS2s', fig=fig_top,
+                                    s2_type_style_id=2)
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:145]:
```suggestion
        fig_top, plot = plot_posS2s(peaks[m_other_s2], label='OS2s', fig=fig_top, s2_type_style_id=2)
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 3]
```diff
@@ -179,15 +178,19 @@ def event_display_interactive(events,
     if plot_record_matrix:
         if st.is_stored(run_id, 'records'):
             # Check if records can be found and load:
-            r = st.get_array(run_id, 'records', time_range=(events[0]['time'], events[0]['endtime']))
+            r = st.get_array(run_id, 'records',
+                             time_range=(events[0]['time'], events[0]['endtime']))
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:182]:
```suggestion
            r = st.get_array(run_id, 'records', time_range=(events[0]['time'], events[0]['endtime']))
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 4]
```diff
@@ -179,15 +178,19 @@ def event_display_interactive(events,
...
         elif st.is_stored(run_id, 'raw_records'):
-            warnings.warn(f'Cannot find records for {run_id}, making them from raw_records instead.')
+            warnings.warn(
+                f'Cannot find records for {run_id}, making them from raw_records instead.')
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:185]:
```suggestion
            warnings.warn(f'Cannot find records for {run_id}, making them from raw_records instead.')
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 5]
```diff
@@ -179,15 +178,19 @@ def event_display_interactive(events,
...
             p = st.get_single_plugin(run_id, 'records')
-            r = st.get_array(run_id, 'raw_records', time_range=(events[0]['time'], events[0]['endtime']))
+            r = st.get_array(run_id, 'raw_records',
+                             time_range=(events[0]['time'], events[0]['endtime']))
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:188]:
```suggestion
            r = st.get_array(run_id, 'raw_records', time_range=(events[0]['time'], events[0]['endtime']))
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 6]
```diff
@@ -199,8 +202,9 @@ def event_display_interactive(events,
         wf, record_points, time_stream = _hvdisp_plot_records_2d(records=r,
                                                                  to_pe=to_pe,
                                                                  t_reference=peaks[0]['time'],
-                                                                 event_range=(waveform.x_range.start,
-                                                                              waveform.x_range.end),
+                                                                 event_range=(
+                                                                 waveform.x_range.start,
+                                                                 waveform.x_range.end),
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:207]:
```suggestion
                                                                 event_range=(waveform.x_range.start,
                                                                              waveform.x_range.end),
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 7]
```diff
@@ -317,7 +321,8 @@ def plot_pmt_arrays_and_positions(top_array_keys,
 
             if pmt_array_type == 'top' and 's2' in k:
                 # In case of the top PMT array we also have to plot the S2 positions:
-                fig, plot = plot_posS2s(signal[k][0], label=labels[k], fig=fig, s2_type_style_id=ind)
+                fig, plot = plot_posS2s(signal[k][0], label=labels[k], fig=fig,
+                                        s2_type_style_id=ind)
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:325]:
```suggestion
                fig, plot = plot_posS2s(signal[k][0], label=labels[k], fig=fig, s2_type_style_id=ind)
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 8]
```diff
@@ -366,10 +371,10 @@ def plot_event(peaks, signal, labels, event, colors, yscale='linear'):
 
     # Get some meaningful x-range limit to 10% left and right extending
     # beyond first last peak, clip at event boundary.
-    length = (end - start) / 10**3
+    length = (end - start) / 10 ** 3
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:374]:
```suggestion
    length = (end - start) / 10**3
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 9]
```diff
@@ -366,10 +371,10 @@ def plot_event(peaks, signal, labels, event, colors, yscale='linear'):
...
 
-    waveform.x_range.start = max(-0.1 * length, (event['time'] - start) / 10**3)
-    waveform.x_range.end = min(1.1 * length, (event['endtime'] - start) / 10**3)
+    waveform.x_range.start = max(-0.1 * length, (event['time'] - start) / 10 ** 3)
+    waveform.x_range.end = min(1.1 * length, (event['endtime'] - start) / 10 ** 3)
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:377]:
```suggestion
    waveform.x_range.start = max(-0.1 * length, (event['time'] - start) / 10**3)
    waveform.x_range.end = min(1.1 * length, (event['endtime'] - start) / 10**3)
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 10]
```diff
@@ -704,7 +709,7 @@ def _make_event_title(event, run_id, width=1600):
     """
     start = event['time']
     date = np.datetime_as_string(start.astype('<M8[ns]'), unit='s')
-    start_ns = start - (start // 10**9) * 10**9
+    start_ns = start - (start // 10 ** 9) * 10 ** 9
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:712]:
```suggestion
    start_ns = start - (start // 10**9) * 10**9
```

File: straxen/analyses/bokeh_waveform_plot.py [hunk 11]
```diff
@@ -718,7 +723,7 @@ def _make_event_title(event, run_id, width=1600):
                              sizing_mode='scale_both',
                              width=width,
                              default_size=width,
-#                              orientation='vertical',
+                             #                              orientation='vertical',
```
Comment (JoranAngevaare) [inline straxen/analyses/bokeh_waveform_plot.py:726]:
```suggestion
                             # orientation='vertical',
```

File: straxen/analyses/holoviews_waveform_display.py [hunk 12]
```diff
@@ -156,7 +156,7 @@ def _hvdisp_plot_records_2d(records,
                             tools=(),
                             hooks=(),
                             plot_library='bokeh',
-                            unit_conversion=10**3):
+                            unit_conversion=10 ** 3):
```
Comment (JoranAngevaare) [inline straxen/analyses/holoviews_waveform_display.py:159]:
```suggestion
                            unit_conversion=10**3):
```

File: straxen/analyses/posrec_comparison.py [hunk 13]
```diff
@@ -21,34 +20,38 @@ def load_corrected_positions(context, run_id, events,
...
-        raise TypeError(f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
 
-    cmt_version = (cmt_version, ) * len(posrec_algos) if isinstance(cmt_version, str) else cmt_version
+    if hasattr(cmt_version, '__len__') and not isinstance(cmt_version, str) and len(
+            cmt_version) != len(posrec_algos):
+        raise TypeError(
+            f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
+
+    cmt_version = (cmt_version,) * len(posrec_algos) if isinstance(cmt_version,
+                                                                   str) else cmt_version
```
Comment (JoranAngevaare) [inline straxen/analyses/posrec_comparison.py:40]:
```suggestion
    cmt_version = (cmt_version, ) * len(posrec_algos) if isinstance(cmt_version, str) else cmt_version
```

File: straxen/analyses/posrec_comparison.py [hunk 14]
```diff
@@ -21,34 +20,38 @@ def load_corrected_positions(context, run_id, events,
...
             raise ValueError(f'CMT is not set? Your fdc config is {fdc_config}') from e
-    
-    if hasattr(cmt_version, '__len__') and not isinstance(cmt_version, str) and len(cmt_version) != len(posrec_algos):
-        raise TypeError(f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
 
-    cmt_version = (cmt_version, ) * len(posrec_algos) if isinstance(cmt_version, str) else cmt_version
+    if hasattr(cmt_version, '__len__') and not isinstance(cmt_version, str) and len(
+            cmt_version) != len(posrec_algos):
+        raise TypeError(
+            f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
```
Comment (JoranAngevaare) [inline straxen/analyses/posrec_comparison.py:37]:
```suggestion
  if hasattr(cmt_version, '__len__') and not isinstance(cmt_version, str) and len(cmt_version) != len(posrec_algos):
        raise TypeError(f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
```

File: straxen/analyses/posrec_comparison.py [hunk 15]
```diff
@@ -21,34 +20,38 @@ def load_corrected_positions(context, run_id, events,
...
+
+    cmt_version = (cmt_version,) * len(posrec_algos) if isinstance(cmt_version,
+                                                                   str) else cmt_version
 
     # Get drift from CMT
-    drift_conf = context.get_single_plugin(run_id, 'event_positions').config.get('electron_drift_velocity')
+    drift_conf = context.get_single_plugin(run_id, 'event_positions').config.get(
+        'electron_drift_velocity')
```
Comment (JoranAngevaare) [inline straxen/analyses/posrec_comparison.py:44]:
```suggestion
    drift_conf = context.get_single_plugin(run_id, 'event_positions').config.get('electron_drift_velocity')
```

File: straxen/analyses/posrec_comparison.py [hunk 16]
```diff
@@ -59,12 +62,13 @@ def load_corrected_positions(context, run_id, events,
             ((f'Interaction angular position (radians) - {algo.upper()} '
               f'posrec algorithm', f'theta_{algo}'),
              np.float32)]
-        
-    dtype += [(('Interaction z-position using mean drift velocity only (cm)', 'z_naive'), np.float32)]
+
+    dtype += [
+        (('Interaction z-position using mean drift velocity only (cm)', 'z_naive'), np.float32)]
```
Comment (JoranAngevaare) [inline straxen/analyses/posrec_comparison.py:67]:
```suggestion
    dtype += [(('Interaction z-position using mean drift velocity only (cm)', 'z_naive'), np.float32)]
```

**Comment by JoranAngevaare** [comment]
my apologies for all this spam due to my pycharm going a bit too enthusiastically through the mini-analyses.

**Comment by WenzDaniel** [comment]
@JoranAngevaare do you mind to bump already strax in your PR to the new version?

**Comment by JoranAngevaare** [comment]
@WenzDaniel as discussed, this isn't required 😄

File: straxen/analyses/quick_checks.py [hunk 17]
```diff
@@ -196,15 +196,17 @@ def plot_energy_spectrum(
         if exposure_kg_sec is not None:
             unit = 'kg_day_kev'
         else:
-            unit = 'events'       
-    
+            unit = 'events'
```
Comment (WenzDaniel) [inline straxen/analyses/quick_checks.py:199]: Not number of events or short '#events' ?

File: straxen/analyses/posrec_comparison.py [hunk 18]
```diff
@@ -21,34 +20,41 @@ def load_corrected_positions(context, run_id, events,
...
         except IndexError as e:
             raise ValueError(f'CMT is not set? Your fdc config is {fdc_config}') from e
-    
-    if hasattr(cmt_version, '__len__') and not isinstance(cmt_version, str) and len(cmt_version) != len(posrec_algos):
-        raise TypeError(f"cmt_version is a list but does not match the posrec_algos ({posrec_algos}) length.")
 
-    cmt_version = (cmt_version, ) * len(posrec_algos) if isinstance(cmt_version, str) else cmt_version
+    if (
+            hasattr(cmt_version, '__len__')
+            and not isinstance(cmt_version, str)
```
Comment (WenzDaniel) [inline straxen/analyses/posrec_comparison.py:36]: Is not it easier to ask if it is a tuple or list?

**Review by WenzDaniel — APPROVED** [review]
Thanks, looks good to me. Can you maybe add either short comments or a doc string for your tests in test_mini_analyses. Sometimes it is not directly clear what exactly you are testing and what the desired outcome should be.

File: straxen/analyses/quick_checks.py (hunk 17 above)
Comment (JoranAngevaare) [reply straxen/analyses/quick_checks.py:199]: as this is only internal I don't think its much of an issue, I actually did not even know we had this function to be honest

File: straxen/analyses/posrec_comparison.py (hunk 18 above)
Comment (JoranAngevaare) [reply straxen/analyses/posrec_comparison.py]: goof point!

**Comment by JoranAngevaare** [comment]
@WenzDaniel , I've added a few docstrings, next time we should really enforce people to write these tests, I don't want to keep having to do these kind of things

**Comment by JoranAngevaare** [comment]
Thanks for the review, much appreciated 👍

File: straxen/analyses/posrec_comparison.py (hunk 18 above)
Comment (JoranAngevaare) [reply straxen/analyses/posrec_comparison.py]: good point

**Comment by WenzDaniel** [comment]
Agree in straxen we have always been a bit more sloppy also for the plugins. Should change it,

**Comment by WenzDaniel** [comment]
Okay this is new :D

**Comment by JoranAngevaare** [comment]
yeah I just added it, I like it 😄

**Pull request merged**